### PR TITLE
Buggeroff-math-fix-and-etc

### PR DESCRIPTION
**Pull request by efrec**
After talking with @Flameink there were clear regressions in builder buggeroff to fix. See https://github.com/beyond-all-reason/Beyond-All-Reason/pull/6057

### Work done

- Fixed area test not being cylindrical (to match blocking)
- Fixed wrong sign when checking if unit's future position is in area
- Better search radius: Per Flameink this needs to be a fixed size, and per Floris' performance optimizations that is doubly preferable. Made that work.
- Better buggeroff radius: Catch very-large units without requiring the buggeroff radius to grow, first.
- Faster updates. Performance seemed good enough but should test.
- Reverted move order change. This was not a general improvement.

Should get a round of comments to be sure we are in the clear.

### CODE REVIEW

**Comment by efrec** [comment]
Builder buggeroff testing with 160 speed unit:

https://github.com/user-attachments/assets/1e88a5f3-7ba8-487c-9b04-76856cdfa74f

Still would like to hear about the buggeroff gadget's many stress tests/edge cases before asking for a merge.

**Comment by Flameink** [comment]
I've tested this out with my usual set of test cases, and it looks like it's working well! The only thing is that the radius increment feels pretty aggressive, I'd cut that down by half.

File: luarules/gadgets/cmd_build_bugger_off.lua [hunk 1]
```diff
@@ -196,22 +202,21 @@ function gadget:GameFrame(frame)
 			-- if there are many units in the way, they may cause a traffic jam and need to clear more room.
 			builderRadiusOffsets[builderID] = builderRadiusOffsets[builderID] + BUGGEROFF_RADIUS_INCREMENT
 
-			for _, interferingUnitID in ipairs(interferingUnits) do
-				if builderID ~= interferingUnitID and not visitedUnits[interferingUnitID] and Spring.GetUnitIsBeingBuilt(interferingUnitID) == false then
+			for _, unitID in ipairs(interferingUnits) do
```
Comment (Flameink) [inline luarules/gadgets/cmd_build_bugger_off.lua]: I feel unitID is kind of vague by itself. Do you think this is an improvement in readability?

Comment (Flameink) [reply luarules/gadgets/cmd_build_bugger_off.lua]: Kind of a nitpick so whatever

Comment (efrec) [reply luarules/gadgets/cmd_build_bugger_off.lua]: I liked this for the brevity but yeah I mentioned in the commit msg that I'd swap it back if anyone complained, so I will.

**Comment by efrec** [comment]
K I think this looks good. We had other notes but they just aren't relevant to this specific PR.

**Review by Flameink — APPROVED** [review]
Looks great to me! Thanks for the fixes. I'll test again with the new changes tonight just in case.

File: luarules/gadgets/cmd_build_bugger_off.lua [hunk 2]
```diff
@@ -36,29 +36,61 @@ for unitDefID, unitDef in pairs(UnitDefs) do
...
+local BUILDER_BUILD_RADIUS  = 200 -- ! varies per-unit
+-- Assume the units are super-fast and medium-sized.
+local SEARCH_RADIUS_OFFSET  = unitSpeedMax + 2 * (Game.squareSize * Game.footprintScale)
+local FAST_UPDATE_FREQUENCY = gameSpeed * 0.5
+local SLOW_UPDATE_FREQUENCY = gameSpeed * 1.5
+local BUGGEROFF_RADIUS_INCREMENT = 2 * Game.squareSize
+-- Move away based on predicted position with lookahead:
+local BUGGEROFF_LOOKAHEAD   = (1/6) * gameSpeed
+-- Limit the buggeroff time to a tortured enough duration:
+local MAX_BUGGEROFF_RADIUS  = BUGGEROFF_RADIUS_INCREMENT * (16 * gameSpeed / SLOW_UPDATE_FREQUENCY) -- => 341 elmos; try to stay < 400
```
Comment (sprunk) [inline luarules/gadgets/cmd_build_bugger_off.lua]: what happens at 400?

Comment (efrec) [reply luarules/gadgets/cmd_build_bugger_off.lua]: Stated

File: luarules/gadgets/cmd_build_bugger_off.lua [hunk 3]
```diff
@@ -36,29 +36,61 @@ for unitDefID, unitDef in pairs(UnitDefs) do
...
+local BUILDER_BUILD_RADIUS  = 200 -- ! varies per-unit
+-- Assume the units are super-fast and medium-sized.
+local SEARCH_RADIUS_OFFSET  = unitSpeedMax + 2 * (Game.squareSize * Game.footprintScale)
+local FAST_UPDATE_FREQUENCY = gameSpeed * 0.5
+local SLOW_UPDATE_FREQUENCY = gameSpeed * 1.5
+local BUGGEROFF_RADIUS_INCREMENT = 2 * Game.squareSize
+-- Move away based on predicted position with lookahead:
+local BUGGEROFF_LOOKAHEAD   = (1/6) * gameSpeed
+-- Limit the buggeroff time to a tortured enough duration:
+local MAX_BUGGEROFF_RADIUS  = BUGGEROFF_RADIUS_INCREMENT * (16 * gameSpeed / SLOW_UPDATE_FREQUENCY) -- => 341 elmos; try to stay < max leash radius (400)
```
Comment (sprunk) [inline luarules/gadgets/cmd_build_bugger_off.lua]: * where does the 16 come from?
* what I meant previously if somebody were to tweak the value in the future, "try to" is a bit too vague to help them make the correct decision vs the tradeoff. If it's definiteively pointless to increase the value beyond (crashes, won't do anything) then be more stern and say so (e.g. "DO NOT go above 400") otherwise specify what generally happens ("try not to because units are otherwise slower to respond" or "because performance" or whatnot).
* ctrl+F "leash" doesn't find anything else in this file, does "max leash radius" refer to `FAST_UPDATE_RADIUS` (in which case consider saying "fast update radius" instead), or some other leash e.g. engine maneuver (in which case consider specifying it explicitly)?

Comment (efrec) [reply luarules/gadgets/cmd_build_bugger_off.lua]: Didn't see this, sorry, I updated this again and caught some of what you mean but not the rest. Thanks.

Comment (sprunk) [reply luarules/gadgets/cmd_build_bugger_off.lua]: Looks like you addressed 2nd and 3rd bullet points. The 1st changed from 16 to 13 and I still don't see where it comes from (comment suggests "time" but I am none the wiser).

**Comment by efrec** [comment]
That last commit lowered the total time to attempt buggeroffs slightly, so @Flameink sorry but if you had any stress tests that only barely worked, we should check them again.

File: luarules/gadgets/cmd_build_bugger_off.lua (hunk 3 above)
Comment (sprunk) [reply luarules/gadgets/cmd_build_bugger_off.lua]: Unless the 13 is actually a tweakable value and it's within a calculation rather than a standalone `local MAX_BUGGER_DURATION = 13` to keep people aware of the 400 limit for the derived value? In that case consider dumping a `-- you can tweak the time` or something like that but otherwise looks good too.

Comment (efrec) [reply luarules/gadgets/cmd_build_bugger_off.lua]: I think an explicit variable is good enough, but time/radius being the same value is more of a deficit in the code that I do think should be addressed, also.
